### PR TITLE
[sdk-metrics] Simplify the state built when an instrument is published

### PR DESCRIPTION
**Pull request by CodeBlanch**
[Most of this first appeared on #5131]

## Changes

* The goal of this work is to simplify the metrics implementation inside the SDK. It removes concepts like "single stream" and "super lists" and instead uses a simple `List<Metric>` for all publishes. There are still optimization(s) for a publish which results in a single `Metric` (the common case).

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)

### CODE REVIEW

**Review by vishweshbankwar — APPROVED** [review]

File: src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs [hunk 1]
```diff
@@ -11,126 +10,33 @@ namespace OpenTelemetry.Metrics;
...
-    {
-        Debug.Assert(metrics.Count == this.count, "The count of metrics to be updated for a CompositeReader must match the number of individual readers.");
-
-        int index = 0;
-        for (var cur = this.Head; cur != null; cur = cur.Next)
-        {
-            var metric = metrics[index];
-            if (metric != null)
+            var innerMetrics = cur.Value.AddMetricWithNoViews(instrument);
+            if (innerMetrics.Count > 0)
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs:20]: Add a `Debug.Assert` here to check that `innerMetrics.Count <= 1`.

File: src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs [hunk 2]
```diff
@@ -11,126 +10,33 @@ namespace OpenTelemetry.Metrics;
...
-            index++;
-        }
+        return metrics;
     }
 
-    internal void CompleteSingleStreamMeasurements(List<Metric?> metrics)
+    internal override List<Metric> AddMetricWithViews(Instrument instrument, List<MetricStreamConfiguration?> metricStreamConfigs)
     {
-        Debug.Assert(metrics.Count == this.count, "The count of metrics to be updated for a CompositeReader must match the number of individual readers.");
+        var metrics = new List<Metric>(this.count);
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs:31]:
```suggestion
        var metrics = new List<Metric>(this.count * metricStreamConfigs.Count);
```

**Review by utpilla — APPROVED** [review]
Left a few suggestions. Great improvement @CodeBlanch !

**Comment by utpilla** [comment]
@CodeBlanch Since this PR is updating a lot of files, could you also run the metrics benchmarks just to ensure that the perf is not impacted due to any non-obvious effects of these changes that might be missed in PR review?

File: src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs:21]: I pushed this and a few other asserts that seemed to be missing.

**Pull request merged**